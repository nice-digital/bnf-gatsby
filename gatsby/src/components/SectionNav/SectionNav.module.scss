@use '@nice-digital/nds-core/scss/core';

.nav {
  background: core.$nds-colour-nice-white;
  border: 1px solid core.$nds-colour-nice-grey-2;
  margin-bottom: core.rem(core.$nds-spacing-large);
  padding: core.rem(core.$nds-spacing-medium);
  position: sticky;
  top: 0;
  z-index: 1;

  &.navReadableMaxWidth {
    max-width: core.$nds-readable-width;
  }

  &.stuck {
    margin-left: -1rem; // TODO: replace with design system gutter vars
    width: calc(100% + 2rem); // TODO: replace with design system gutter vars
  }

  @media print {
    display: none;
  }
}

.heading {
  @include core.nds-h4;
  margin-top: 0;
}

.linkList {
  list-style: none;
  margin: 0;
  max-width: none;
  padding: 0;

  li {
    line-height: 1.3;
    margin-bottom: 0;
    padding: core.rem(core.$nds-spacing-x-small) 0;
  }
}

@include core.mq($from: md) {
  .linkList {
    column-gap: core.rem(core.$nds-spacing-large);
    columns: 2;

    li {
      break-inside: avoid;
    }
  }
}
