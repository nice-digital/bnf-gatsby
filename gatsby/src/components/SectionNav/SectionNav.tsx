import { FC } from "react";
import striptags from "striptags";

import styles from "./SectionNav.module.scss";

export interface SectionNavProps {
	sections: {
		id: string;
		title: string;
	}[];
}

export const SectionNav: FC<SectionNavProps> = ({ sections }) => (
	<nav aria-labelledby="navigate-to-section" className={styles.nav}>
		<h2 id="navigate-to-section" className={styles.heading}>
			Navigate to section
		</h2>
		<ol
			aria-label="Jump links to sections on this page"
			className={styles.linkList}
		>
			{sections.map((section) => (
				<li key={section.id}>
					<a href={`#${section.id}`}>{striptags(section.title)}</a>
				</li>
			))}
		</ol>
	</nav>
);
