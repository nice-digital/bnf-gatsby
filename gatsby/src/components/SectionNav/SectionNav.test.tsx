import { render, screen } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import { renderToString } from "react-dom/server";

import { SectionNav } from "./SectionNav";

const defaultProps = {
	sections: [
		{ title: "test-1", id: "wound-type-1889862293" },
		{ title: "test-2", id: "wound-type-1889862294" },
	],
};

describe("SectionNav", () => {
	it("should match snapshot server side", () => {
		const wrapper = document.createElement("div");
		wrapper.innerHTML = renderToString(<SectionNav {...defaultProps} />);

		expect(wrapper).toMatchSnapshot();
	});

	it("should match snapshot client side", () => {
		const { container } = render(<SectionNav {...defaultProps} />);
		expect(container).toMatchSnapshot();
	});

	describe("Button", () => {
		it("should render a button with an accessible label to show sections", () => {
			render(<SectionNav {...defaultProps} />);
			expect(
				screen.getByRole("button", {
					name: `Show Navigate to section`,
				})
			).toBeInTheDocument();
		});

		it("should not be expanded by default", () => {
			render(<SectionNav {...defaultProps} />);
			expect(
				screen.getByRole("button", {
					name: `Show Navigate to section`,
				})
			).toHaveAttribute("aria-expanded", "false");
		});

		it("should expand on click", () => {
			render(<SectionNav {...defaultProps} />);
			const button = screen.getByRole("button", {
				name: `Show Navigate to section`,
			});
			userEvent.click(button);
			screen.debug();
			expect(button).toHaveAttribute("aria-expanded", "true");
		});

		it("should render an icon with an expanded class on click", () => {
			render(<SectionNav {...defaultProps} />);
			const button = screen.getByRole("button", {
				name: `Show Navigate to section`,
			});

			// eslint-disable-next-line testing-library/no-node-access
			const svg = button.querySelector("svg");
			expect(svg?.classList.value).toEqual("icon");
			userEvent.click(button);
			expect(svg).toHaveClass("icon iconExpanded");
		});
	});

	describe("Link list", () => {
		it("should render links when button is clicked", () => {
			render(<SectionNav {...defaultProps} />);
			const button = screen.getByRole("button", {
				name: `Show Navigate to section`,
			});
			userEvent.click(button);
			expect(screen.getByRole("link", { name: "test-1" })).toBeInTheDocument();
			expect(screen.getByRole("link", { name: "test-2" })).toBeInTheDocument();
		});

		it("should render an aria labelled list in specific order by dom snapshot", () => {
			render(<SectionNav {...defaultProps} />);
			const button = screen.getByRole("button", {
				name: `Show Navigate to section`,
			});

			userEvent.click(button);
			const list = screen.getByRole("list");
			expect(list).toMatchInlineSnapshot(`
			<ol
			  aria-label="Jump links to sections on this page"
			  class="linkList"
			>
			  <li>
			    <a
			      href="#wound-type-1889862293"
			    >
			      test-1
			    </a>
			  </li>
			  <li>
			    <a
			      href="#wound-type-1889862294"
			    >
			      test-2
			    </a>
			  </li>
			</ol>
		`);
		});

		it("should render a list item for each section when button is clicked", () => {
			render(<SectionNav {...defaultProps} />);
			const button = screen.getByRole("button", {
				name: `Show Navigate to section`,
			});
			expect(screen.queryAllByRole("listitem")).toHaveLength(0);
			userEvent.click(button);
			expect(screen.queryAllByRole("listitem")).toHaveLength(2);
		});
	});
});
