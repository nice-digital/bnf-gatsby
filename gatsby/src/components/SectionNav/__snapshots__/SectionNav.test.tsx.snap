// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SectionNav should match snapshot client side 1`] = `
<div>
  <nav
    aria-labelledby="navigate-to-section"
    class="nav"
  >
    <h2
      class="heading"
      id="navigate-to-section"
    >
      <span>
        Navigate to section
      </span>
    </h2>
    <ol
      aria-label="Jump links to sections on this page"
      class="linkList"
    >
      <li>
        <a
          href="#wound-type-1889862293"
        >
          test-1
        </a>
      </li>
      <li>
        <a
          href="#wound-type-1889862294"
        >
          test-2
        </a>
      </li>
    </ol>
  </nav>
</div>
`;

exports[`SectionNav should match snapshot server side 1`] = `
<div>
  <nav
    aria-labelledby="navigate-to-section"
    class="nav"
  >
    <h2
      class="heading"
      id="navigate-to-section"
    >
      <span>
        Navigate to section
      </span>
    </h2>
    <ol
      aria-label="Jump links to sections on this page"
      class="linkList"
    >
      <li>
        <a
          href="#wound-type-1889862293"
        >
          test-1
        </a>
      </li>
      <li>
        <a
          href="#wound-type-1889862294"
        >
          test-2
        </a>
      </li>
    </ol>
  </nav>
</div>
`;
